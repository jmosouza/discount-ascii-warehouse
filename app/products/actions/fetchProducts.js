/* global fetch, API_HOST, API_PER_PAGE */

import ndjsonStream from 'can-ndjson-stream';

export const fetchProductsSuccess = product => ({
  type: 'PRODUCTS::FETCH_SUCCEEDED',
  product,
});

export const fetchProductsNoMore = () => ({
  type: 'PRODUCTS::NO_MORE',
});

// Request products, read NDJSON stream, append products as they come through
export default () => (dispatch, getState) => {
  // The state should know when the API returns no more products.
  // Set this to false when the API returns at least 1 product.
  let noMoreProducts = true;

  const { page, sort } = getState().products;
  const skip = page * API_PER_PAGE;
  fetch(`${API_HOST}/api/products?skip=${skip}&sort=${sort}&limit=${API_PER_PAGE}`)
    .then(data => ndjsonStream(data.body))
    .then((stream) => {
      const streamReader = stream.getReader();
      const readHandler = (result) => {
        if (result.done) {
          if (noMoreProducts) dispatch(fetchProductsNoMore());
          return;
        }

        // There is at least 1 product
        noMoreProducts = false;

        dispatch(fetchProductsSuccess(result.value));
        streamReader.read().then(readHandler);
      };
      streamReader.read().then(readHandler);
    });
};
