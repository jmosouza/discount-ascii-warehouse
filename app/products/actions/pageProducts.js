export default page => ({
  type: 'PRODUCTS::PAGE',
  page,
});
