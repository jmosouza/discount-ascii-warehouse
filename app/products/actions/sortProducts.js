export default sort => ({
  type: 'PRODUCTS::SORT',
  sort,
});
