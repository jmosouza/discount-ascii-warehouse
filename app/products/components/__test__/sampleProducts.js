export default [
  {
    id: '0-kmg1to7ttda', size: 35, price: 114, face: '( .-. )', date: 'Tue Oct 24 2017 12:27:17 GMT-0300 (-03)',
  },
  {
    id: '1-uq52ro7hh3a', size: 15, price: 74, face: '( .o.)', date: 'Tue Oct 24 2017 09:56:25 GMT-0300 (-03)',
  },
  {
    id: '2-kixa14dmqaa', size: 25, price: 700, face: '( `·´ )', date: 'Sat Oct 21 2017 21:40:50 GMT-0300 (-03)',
  },
  {
    id: '3-53gsauxs2tn', size: 18, price: 950, face: '( ° ͜ ʖ °)', date: 'Fri Oct 20 2017 14:49:26 GMT-0300 (-03)',
  },
  {
    id: '4-p2ao8104bd9', size: 35, price: 464, face: '( ͡° ͜ʖ ͡°)', date: 'Fri Oct 20 2017 17:13:23 GMT-0300 (-03)',
  },
  {
    id: '5-eba70ah0kp6', size: 28, price: 917, face: '( ⚆ _ ⚆ )', date: 'Sat Oct 28 2017 12:31:50 GMT-0300 (-03)',
  },
  {
    id: '6-vxb0xq0hcom', size: 16, price: 205, face: '( ︶︿︶)', date: 'Wed Oct 25 2017 15:43:38 GMT-0300 (-03)',
  },
  {
    id: '7-0bepqhppuz5j', size: 37, price: 711, face: '( ﾟヮﾟ)', date: 'Tue Oct 17 2017 23:14:26 GMT-0300 (-03)',
  },
  {
    id: '8-d44oc6ofo7m', size: 17, price: 17, face: '(\\/)(°,,,°)(\\/)', date: 'Thu Oct 26 2017 17:30:00 GMT-0300 (-03)',
  },
  {
    id: '9-3krjuzvy29w', size: 25, price: 623, face: '(¬_¬)', date: 'Tue Oct 17 2017 07:04:19 GMT-0300 (-03)',
  },
];
