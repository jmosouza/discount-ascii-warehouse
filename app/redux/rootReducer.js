import { combineReducers } from 'redux';
import products from '../products/reducers/productsReducer';

export default combineReducers({
  products,
});
