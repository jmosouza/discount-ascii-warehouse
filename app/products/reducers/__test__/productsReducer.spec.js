/* global test, expect */

import rootReducer from '../../../redux/rootReducer';
import { fetchProductsSuccess } from '../../actions/fetchProducts';
import sampleProducts from '../../components/__test__/sampleProducts';

[
  { type: 'UNKNOWN ACTION' },
  fetchProductsSuccess(sampleProducts[0]),
].forEach((action) => {
  test(`Dispatch ${action.type}`, () => {
    expect(rootReducer(undefined, action)).toMatchSnapshot();
  });
});
