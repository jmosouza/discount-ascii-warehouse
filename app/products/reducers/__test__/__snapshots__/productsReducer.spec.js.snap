// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dispatch PRODUCTS::FETCH_SUCCEEDED 1`] = `
Object {
  "products": Object {
    "all": Object {
      "0-kmg1to7ttda": Object {
        "date": "Tue Oct 24 2017 12:27:17 GMT-0300 (-03)",
        "face": "( .-. )",
        "id": "0-kmg1to7ttda",
        "price": 114,
        "size": 35,
      },
    },
    "hasMore": true,
    "page": 0,
    "sort": "id",
  },
}
`;

exports[`Dispatch PRODUCTS::NO_MORE 1`] = `
Object {
  "products": Object {
    "all": Object {},
    "hasMore": false,
    "page": 0,
    "sort": "id",
  },
}
`;

exports[`Dispatch UNKNOWN ACTION 1`] = `
Object {
  "products": Object {
    "all": Object {},
    "hasMore": true,
    "page": 0,
    "sort": "id",
  },
}
`;
